src/ipfs/daemon.rs

```diff
@@ -0,0 +1,221 @@
+
+use std::collections::HashMap;
+use std::process::Command;
+use std::thread;
+use std::time::{SystemTime, Duration};
+use std::path::Path;
+
+use anyhow::{bail, Result};
+use async_trait::async_trait;
+use colored::Colorize;
+use futures::executor::block_on;
+use graceful::SignalGuard;
+use ipfs_api_backend_hyper::{TryFromUri, IpfsClient, IpfsApi, LoggingLevel, Logger};
+use serde_json::Value;
+use tokio::runtime::Runtime;
+
+use crate::ipfs::Ipfs;
+use crate::utils::config::{IPFS_ADDR, IPFS_API_PORT, IPFS_EXE, IPFS_SLEEP_LENGTH, IPFS_BOOT_TIME_OUT};
+
+/// This struct is a wrapper for the information needed to point the IPFS daemon at a diffrent address
+/// than the default.
+#[derive(Clone, Debug)]
+pub struct IpfsConnInfo {
+    pub address:String,
+    pub port:u16
+}
+
+pub struct IpfsDaemon {
+    conn_info: IpfsConnInfo,
+    client: IpfsClient,
+    tokio: Runtime
+}
+
+impl IpfsDaemon {
+    /// This method launches the IPFS daemon on the port/address given when the the instance was
+    /// created. During this proccess it also creates a thread that listens for a shutdown signal
+    /// and stops the IPFS daemon gracefully if the signal is given.
+    pub async fn launch(&self) -> Result<()>{
+        //launch the daemon
+        let api_addr = format!("/ip4/{}/tcp/{}", self.conn_info.address, self.conn_info.port);
+        println!("Launhing IPFS...");
+        Command::new(IPFS_EXE)
+            .arg("--api")
+            .arg(&api_addr)
+            .arg("daemon")
+            .spawn()
+            .unwrap_or_else(|e| panic!("Failed to start IPFS daemon: {}\n This error may be because the Kubo binary is not on your PATH.", e));
+
+        //Wait ipfs to ready
+        println!("Waiting for IPFS to ready..");
+        self.await_ready().await?;
+
+        //Reduce log level for IPFS
+        self.tokio.block_on(async {
+            self.client.log_level(Logger::All, LoggingLevel::Error).await
+        })?;
+        
+        //setup gracefull shutdown
```

Comment:
```suggestion
        // Setup graceful shutdown
```
Noticing that the comment style for `//` and `///` comments is a bit different. 

The `//` ones don't have a space after the `//`. What would you think about making these consistent with the `///` comments by adding spaces?